cli/cmd/proxy/list/command.go

```diff
@@ -0,0 +1,218 @@
+		Name:    "all-namespaces",
+		Target:  &c.flagAllNamespaces,
+		Default: false,
+		Usage:   "List pods in all namespaces.",
+		Aliases: []string{"A"},
+	})
+
+	f = c.set.NewSet("Global Options")
+	f.StringVar(&flag.StringVar{
+		Name:    "kubeconfig",
+		Aliases: []string{"c"},
+		Target:  &c.flagKubeConfig,
+		Default: "",
+		Usage:   "Set the path to kubeconfig file.",
+	})
+	f.StringVar(&flag.StringVar{
+		Name:    "context",
+		Target:  &c.flagKubeContext,
+		Default: "",
+		Usage:   "Set the Kubernetes context to use.",
+	})
+
+	c.help = c.set.Help()
+	c.Init()
+}
+
+// Run executes the list command.
+func (c *ListCommand) Run(args []string) int {
+	c.once.Do(c.init)
+	c.Log.ResetNamed("list")
+	defer common.CloseWithError(c.BaseCommand)
+
+	// Parse the command line flags.
+	if err := c.set.Parse(args); err != nil {
+		c.UI.Output("Error parsing arguments:\n%v", err.Error())
+		return 1
+	}
+
+	// Validate the command line flags.
+	if err := c.validateFlags(); err != nil {
+		c.UI.Output("Invalid argument:\n%v", err.Error())
+		return 1
+	}
+
+	if c.kubernetes == nil {
+		if err := c.initKubernetes(); err != nil {
+			c.UI.Output("Error initializing Kubernetes client", err.Error())
+			return 1
+		}
+	}
+
+	if c.flagAllNamespaces {
+		c.flagNamespace = ""
+	}
+
+	pods, err := c.fetchPods()
+	if err != nil {
+		c.UI.Output("Error fetching pods:\n%v", err, terminal.WithErrorStyle())
+		return 1
+	}
+
+	c.output(pods)
+	return 0
+}
+
+// Help returns a description of the command and how it is used.
+func (c *ListCommand) Help() string {
+	c.once.Do(c.init)
+	return c.Synopsis() + "\n\nUsage: consul-k8s proxy list [flags]\n\n" + c.help
+}
+
+// Synopsis returns a one-line command summary.
+func (c *ListCommand) Synopsis() string {
+	return "List all Pods running Consul proxies."
+}
+
+// validateFlags ensures that the flags passed in by the can be used.
+func (c *ListCommand) validateFlags() error {
+	if len(c.set.Args()) > 0 {
+		return errors.New("should have no non-flag arguments")
+	}
+	if !common.IsValidLabel(c.flagNamespace) {
+		return fmt.Errorf("'%s' is an invalid namespace. Namespaces follow the RFC 1123 label convention and must "+
+			"consist of a lower case alphanumeric character or '-' and must start/end with an alphanumeric character", c.flagNamespace)
+	}
+	return nil
+}
+
+// initKubernetes initializes the Kubernetes client with the users config.
+func (c *ListCommand) initKubernetes() error {
+	settings := helmCLI.New()
+
+	restConfig, err := settings.RESTClientGetter().ToRESTConfig()
+	if err != nil {
+		return fmt.Errorf("error retrieving Kubernetes authentication %v", err)
+	}
+	c.kubernetes, err = kubernetes.NewForConfig(restConfig)
+
+	return err
+}
+
+// fetchPods fetches all pods in flagNamespace which run Consul proxies.
+func (c *ListCommand) fetchPods() ([]v1.Pod, error) {
+	var pods []v1.Pod
+
+	gatewaypods, err := c.kubernetes.CoreV1().Pods(c.flagNamespace).List(c.Ctx, metav1.ListOptions{
```

Comment:
I feel like using just this LabelSelector could bring in other applications that may not have proxies attached.
We also have the following Labels that get attached to the deployments:
```
  selector:
    matchLabels:  
       app: {{ template "consul.name" . }} 
       chart: {{ template "consul.chart" . }}
       release: {{ .Release.Name }}
       component: mesh-gateway
```

```
consul-mesh-gateway-f6d566cd8-x75bn            2/2     Running   0               3d17h   app=consul,chart=consul-helm,component=mesh-gateway,pod-template-hash=f6d566cd8,release=consul
```

Since we only support installing one Consul installation per k8s cluster does it make sense to read the helm secret to figure out the release name and use that to be more explicit on the Selectors throughout the entire `fetchPods` function?

Reply:
Yes, that sounds possible. I would need to fetch the latest secret created by Helm. I see the concern though. A team could have their own code in a pod that has the `component: mesh-gateway` label.

Reply:
I will add this as a bug to fix before the final release.

Reply:
Here is the bug to track this: https://app.asana.com/0/1200500259727746/1202437872167459/f